File: Linear Control Course/Chapter8/Lesson4/Chapter8_Lesson4.m
% Plant and PI controller
s = tf('s');
G = 5 / (s + 3);          % first-order plant
Kp = 1;
Ki = 4;
C = Kp + Ki / s;          % PI controller

L = C * G;
T = feedback(L, 1);       % closed-loop from r to y
S = 1 - T;                % sensitivity from r to e

% Static error constants
Kp_static = dcgain(L);      % position constant
Kv_static = dcgain(s * L);  % velocity constant

fprintf('Kp (static) = %g\n', Kp_static);
fprintf('Kv (static) = %g\n', Kv_static);
fprintf('Predicted ramp error (unit ramp) = %g\n', 1 / Kv_static);

% Step response (zero steady-state error)
figure;
step(T);
title('Step response with PI controller');

% Ramp response via lsim
t = 0:0.01:10;
r_ramp = t;   % unit-slope ramp
[y_ramp, t_out] = lsim(T, r_ramp, t);

figure;
plot(t_out, r_ramp, 'LineWidth', 1.5); hold on;
plot(t_out, y_ramp, 'LineWidth', 1.5);
grid on;
xlabel('Time (s)');
ylabel('Position');
legend('r(t) = ramp', 'y(t)');
title('Ramp tracking with PI controller');

% Simulink notes:
% A corresponding Simulink model uses:
%  - Sum block (r - y)
%  - Discrete or continuous PI controller block
%  - Transfer Fcn block for G(s)
%  - Scope blocks to view y(t) and e(t)

File: Linear_Control_Course/Chapter8/Lesson4/Chapter8_Lesson4.py
import numpy as np
import control as ct
import matplotlib.pyplot as plt

# Plant model: approximated joint actuator (continuous time)
s = ct.TransferFunction.s
G = 5 / (s + 3)

# PI controller from the analytic design
Kp = 1.0
Ki = 4.0
C = Kp + Ki / s  # C(s) = Kp + Ki/s

L = C * G               # loop transfer function
T = ct.feedback(L, 1)   # closed-loop from r to y
S = 1 - T               # sensitivity (from r to e)

# Static error constants (step and ramp)
# Kp_static = lim_{s->0} L(s)
Kp_static = ct.dcgain(L)  # dcgain gives G(0) for proper systems

# Kv = lim_{s->0} s L(s)
Kv_tf = s * L
Kv_static = ct.dcgain(Kv_tf)

print("Kp (static) =", Kp_static)
print("Kv (static) =", Kv_static)
print("Predicted ramp error (unit ramp) =", 1.0 / Kv_static)

# Numerical simulation for step and ramp
t = np.linspace(0, 10, 1000)

# Step response (unit step)
t_step, y_step = ct.step_response(T, T=t)
e_step = 1.0 - y_step

# Ramp response: r(t) = t for t >= 0
r_ramp = t
t_ramp, y_ramp, x_ramp = ct.forced_response(T, T=t, U=r_ramp)
e_ramp = r_ramp - y_ramp

print("Approximate steady-state step error =", e_step[-1])
print("Approximate steady-state ramp error =", e_ramp[-1])

# Plot ramp tracking
plt.figure()
plt.plot(t_ramp, r_ramp, label="r(t) = ramp")
plt.plot(t_ramp, y_ramp, label="y(t)")
plt.xlabel("Time [s]")
plt.ylabel("Position")
plt.legend()
plt.title("Ramp tracking with PI controller")
plt.grid(True)
plt.show()
